### PR TITLE
Unreal: decouple UIs to separate process

### PR DESCRIPTION
**Pull request by simonebarbieri**
## Brief description
Run unreal commands from OP through websocket communication.

## Description
UE has the ability to have its python to be called remotely. This PR allow us to run all UIs as separete processes that are comunicating with UE. We'll no longer need to install PySide2 to Unreal.

### CODE REVIEW

File: openpype/hosts/unreal/api/launch_script.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+
+def main(launch_args):
+
+    # Be sure server won't crash at any moment but just print traceback
+    sys.excepthook = safe_excepthook
+
+    # Create QtApplication for tools
+    # - QApplicaiton is also main thread/event loop of the server
+    qt_app = QtWidgets.QApplication([])
```
Comment (iLLiCiTiT) [inline openpype/hosts/unreal/api/launch_script.py:32]: You can use `get_openpype_qt_app` from `openpype.tools.utils`.

File: openpype/hosts/unreal/api/plugin.py [hunk 2]
```diff
@@ -242,6 +221,24 @@ def get_pre_create_attr_defs(self):
...
+class UnrealBaseLoader(LoaderPlugin):
+    """Base class for Unreal loader plugins."""
+    root = "/Game/Ayon"
+    suffix = "_CON"
+
+    def update(self, container, representation):
+        asset_dir = container["namespace"]
+        container_name = container['objectName']
+
+        data = {
```
Comment (iLLiCiTiT) [inline openpype/hosts/unreal/api/plugin.py:234]: This is optional suggestion, to change the keys to `representation_id` and `version_id`.

**Review by Minkiu — APPROVED** [review]
Haven't tested it, but code looks good, there are some docstrings missing here and ther though.

And was wondering (in another PR) would be worth wrapping all the functions we pass to `send_request` the same way we do with `containerise`.

**Comment by moonyuet** [comment]
launcher works as expected to launch Unreal with AYON 
image

But the Unreal can't open the ayon publisher as it hits the Attribute Error.
```
LogPython: Warning: Ayon: showing tools popup
LogScript: Error: Script Msg: Traceback (most recent call last):
  File "C:/Program Files/Epic Games/UE_5.1/Engine/Plugins/Marketplace/Ayon/Content/Python/init_unreal.py", line 24, in RunInPython_Popup
    ayon_host.show_tools_popup()
AttributeError: 'UnrealHost' object has no attribute 'show_tools_popup'
LogScript: Error: Script call stack:
    /Engine/PythonTypes.AyonIntegration.RunInPython_Popup
```
The detail log command from UE shown below: 
[Unreal_LOG.txt](https://github.com/ynput/OpenPype/files/12713601/Unreal_LOG.txt)

**Review by antirotor — CHANGES_REQUESTED** [review]
So, I've been testing it in **UE 5.3** (so I've made some changes in ynput/ayon-unreal-plugin). Anyway, I've hit few issues:

## Publisher
### Creating camera
When running camera creator:

```
Traceback (most recent call last):
  File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\pipeline\create\context.py", line 2030, in _create_with_unified_error
    result = creator.create(*args, **kwargs)
  File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\hosts\unreal\plugins\create\create_camera.py", line 27, in create
    instance_data["level"] = send_request("get_editor_world")
  File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\hosts\unreal\api\pipeline.py", line 162, in send_request
    return ret_value.get("return")
AttributeError: 'str' object has no attribute 'get'
```

### Creating Layout
When running layout creator:
```
DEBUG:openpype.hosts.unreal.api.communication_server:Sending request to client localhost (get_editor_world, {}) id: 36
WARNING:CreateContext:Failed to run Creator with identifier "io.ayon.creators.unreal.layout". Creator error: 'str' object has no attribute 'get'
```

### Creating Look
When running Look creator:
```
WARNING:CreateContext:Failed to run Creator with identifier "io.ayon.creators.unreal.look".
Traceback (most recent call last):
  File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\pipeline\create\context.py", line 2030, in _create_with_unified_error
    result = creator.create(*args, **kwargs)
  File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\hosts\unreal\plugins\create\create_look.py", line 40, in create
    pre_create_data["members"] = send_request(
  File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\hosts\unreal\api\pipeline.py", line 160, in send_request
    communicator.send_request(request, params)
  File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\hosts\unreal\api\communication_server.py", line 605, in send_request
    return self.websocket_rpc.send_request(
  File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\hosts\unreal\api\communication_server.py", line 291, in send_request
    raise Exception("Error happened: {}".format(error))
Exception: Error happened: {'code': -32000, 'message': 'Python Execution in Unreal Failed!', 'data': ''}
```

### Creating Renders:

This happens probably because of json serialization:
```
WARNING:CreateContext:Failed to run Creator with identifier "io.ayon.creators.unreal.render".
Traceback (most recent call last):
  File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\pipeline\create\context.py", line 2030, in _create_with_unified_error
    result = creator.create(*args, **kwargs)
  File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\hosts\unreal\plugins\create\create_render.py", line 97, in create
    self.create_with_new_sequence(
  File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\hosts\unreal\plugins\create\create_render.py", line 50, in create_with_new_sequence
    master_lvl, sequence, seq_data = send_request(
  File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\hosts\unreal\api\pipeline.py", line 159, in send_request
    if ret_value := ast.literal_eval(
  File "C:\Users\annat\.pyenv\pyenv-win\versions\3.9.13\lib\ast.py", line 62, in literal_eval
    node_or_string = parse(node_or_string, mode='eval')
  File "C:\Users\annat\.pyenv\pyenv-win\versions\3.9.13\lib\ast.py", line 50, in parse
    return compile(source, filename, mode, flags,
  File "<unknown>", line 1
    ('/Game/Ayon/Sequences/renderGenericMain/renderGenericMain_MasterLevel', '/Game/Ayon/Sequences/renderGenericMain/renderGenericMain.renderGenericMain', {'sequence': <Object '/Game/Ayon/Sequences/renderGenericMain/renderGenericMain.renderGenericMain' (0x00000B2D4B1E8700) Class 'LevelSequence'>, 'output': 'renderGenericMain', 'frame_range': (0, 150)})

                                                ^
SyntaxError: invalid syntax
```

## Scene Inventory
This doesn't work - it won't list any loaded instances (`ls()` returning empty list). This might be caused by merging changes in  ynput/ayon-unreal-plugin though.

## Loading
Loading works, as far as I could test it - I would just add timestamp to server console.

**TODO:**

- We need to test it in AYON
- Resolve conflicts (merge 3.17 in)

**Comment by antirotor** [comment]
This will be implemented only for AYON.

**Pull request closed without merging**